mutating-admission-controller/main.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/api/admission/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	port = "8080"
+)
+
+var (
+	tlscert, tlskey string
+)
+
+type patchOperation struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+type admitFunc func(v1.AdmissionReview) *v1.AdmissionResponse
+
+func toAdmissionResponse(err error) *v1.AdmissionResponse {
+	return &v1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+		},
+	}
+}
+
+func serveMutateServiceAccount(w http.ResponseWriter, r *http.Request) {
+	serve(w, r, mutateServiceAccount)
+}
+
+func serve(w http.ResponseWriter, r *http.Request, admit admitFunc) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+	if len(body) == 0 {
+		glog.Error("empty body")
+		http.Error(w, "empty body", http.StatusBadRequest)
+		return
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		glog.Errorf("Content-Type=%s, expect application/json", contentType)
+		http.Error(w, "invalid Content-Type, expect `application/json`", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	// The AdmissionReview that was sent to the webhook
+	requestedAdmissionReview := v1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := v1.AdmissionReview{}
+
+	deserializer := scheme.Codecs.UniversalDeserializer()
+
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		glog.Error(err)
+		responseAdmissionReview.Response = toAdmissionResponse(err)
+	} else {
+		responseAdmissionReview.Response = admit(requestedAdmissionReview)
+	}
+
+	responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	responseAdmissionReview.APIVersion = "admission.k8s.io/v1"
+	responseAdmissionReview.Kind = "AdmissionReview"
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		glog.Error(err)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		glog.Error(err)
+	}
+}
+
+func mutateServiceAccount(ar v1.AdmissionReview) *v1.AdmissionResponse {
+	req := ar.Request
+
+	glog.Infof("AdmissionReview for Kind=%v, Namespace=%v Name=%v UID=%v patchOperation=%v UserInfo=%v",
+		req.Kind, req.Namespace, req.Name, req.UID, req.Operation, req.UserInfo)
+
+	if req.Kind.Kind == "ServiceAccount" && req.Name == "default" {
```

Comment:
If we check for default names, this mean that if user OR component creates any service account, it will still get the token automatically mounted. In many cases, ServiceAccount is only required to allow pod to use specific pod security policy and not to talk to Kubernetes API, so in my opinion, this could be improved.